### DISCUSSION TITLE
How to calculate changes in feature importance over time for a given model?

### DISCUSSION THREAD

**Discussion by abeboparebop** [shap/shap#1487] [General]
Suppose I have a model trained on 2017 data, which is used to make predictions on 2018 and 2019 data. I observe that the scores are different in some way between 2018 and 2019 (e.g. higher predictions on average, or worse performance on standard ML metrics), and I want to use the KernelExplainer to understand which features are driving the change. I would use these importances to make a claim like: "features X, Y, and Z had much higher importance during 2019 than during 2018, so these are the features that are driving the change in predictions."

For calculating feature importances during 2018 and 2019, which time periods should I draw my background sample from? I can see various possibilities, and probably there are others I'm overlooking:

1. For feature importances in 2018, I should use a background sample from 2018, and for 2019 I should use a sample from 2019.
2. For feature importances in both years I should use the same background sample, drawn from both 2018 and 2019.
3. For feature importances in both years I should use the same background sample, drawn from only 2018 (because 2018 is the baseline for comparison).
4. For feature importances in both years I should use the same background sample, drawn from 2017 (because this represents the true baseline feature distribution that the model was trained on).

My question is practical, but perhaps if I had a better conceptual understanding the answer would be obvious, so I'm interested in that side of things as well. :)

**Comment by slundberg** [comment]
Good questions! Since SHAP values always explain why a feature is important relative to an alternative, that alternative (background distribution) is important to keep in mind. While there are many creative ways to use the background distribution (like comparing to samples), typically we choose the background distribution to represent a prior we have over the inputs. So it is what we would expect when we don't know anything about an input value. This leads to the intuitive notion of "hiding/masking" inputs by integrating over the background distribution. In psychology this kind of hiding is called a "subtractive counterfactual" and is used by humans when they explain things.

So in your case I would ask what it would mean to "hide" something. Normally it would be to integrate over the training data set, which would be different as you point out between models from different years. However this could create a difference between the explanations based on their backgrounds being different, so I think for the sake of model comparison I would pick a single background dataset for both models. One way to think this about this is to consider that both models could really just be viewed as a single model, where the single model just uses one of the sub-models depending on the year. That would argue for a joint 2018-2019 background. However, it might also be nice to have an anchor in the past and just use 2018...you just need to remember the alternative you are comparing against when interpreting the explanations.
